vmm/src/device_manager.rs

```diff
@@ -491,6 +491,26 @@ impl DeviceManager {
         })
     }
 
+    pub fn serde_devices(&self) -> String {
+        let devices = self.address_manager.io_bus.devices.read().unwrap();
+        let mut json = String::new();
+
+        for (_range, device) in devices.iter() {
+            let mut guard = device.lock().unwrap();
+
+            if let Some(dev) = guard.as_any().downcast_ref::<devices::legacy::Serial>() {
+                println!("Serial device.");
+                let serial = crate::migration::device_states::serde_serial(dev);
+                json.push_str(&serial);
+            }
+        }
```

Comment:
Downcasting means we'll have to repeat this operation for each and every concrete type that we want to migrate: serial, cmos, virtio-net, vhost-user-blk, etc. I was hoping that we could all `serde_json()` directly on each device from `io_bus`.
Instead, the `serde_device()` routine is essentially saying: Go over each device and *try* to serialize its state. This is equivalent to say: Go over each device that can migrate and call it's `snapshot()` implementation.

What I'm saying here is that you're implementing the `Migratable` trait mentioned in PR #444, except that:

1. You'll have to downcast each and every BusDevice to all the concrete type you want to migrate.
2. You're going to carry device specific snapshot implementations in the migration crate.

I was hoping this PR to magically serialize all BusDevice, Cpu, etc, without having to manually downcast and call ad-hoc serializing functions. If we can't do that, then a `Migratable` trait is more idiomatic and provides a cleaner separation of duties (The VMM does the snapshotting management, each component handles its actual snapshot).